pkg/dyninst/symdb/func_name.go

```diff
@@ -0,0 +1,141 @@
+	// parse correctly if there is a type name and no package name. Hopefully that
+	// doesn't exist.
+	//   - note that we don't include the optional '*' in the capture group. We
+	//     don't differentiate between pointer and value receivers.
+	//   - note the lazy capture group for the type name: .*?. It's lazy because
+	//     otherwise it also captures the following ')'
+	//   - the function name: (.*)
+	parseFuncNameRE = regexp.MustCompile(`^((?P<pkg>(.*/)?.*?)\.)?(\(?\*?(?P<typ>.*?)\)?\.)?(?P<name>.*)$`)
+	pkgIdx          = parseFuncNameRE.SubexpIndex("pkg")
+	typIdx          = parseFuncNameRE.SubexpIndex("typ")
+	nameIdx         = parseFuncNameRE.SubexpIndex("name")
+
+	// Recognize anonymous functions declared inside a function that is not a method.
+	// They look like:
+	// github.com/.../pkg.myFunc.func1
+	// internal/bytealg.init.0
+	parseAnonymousFuncNameRE = regexp.MustCompile(`^((?P<pkg>(.*/)?\w*?)\.)?(?P<func>(\w*?\.)?((func|gowrap|(\d+)).*))$`)
+	anonPkgIdx               = parseFuncNameRE.SubexpIndex("pkg")
+	anonNameIdx              = parseFuncNameRE.SubexpIndex("name")
+
+	// Recognize funky functions corresponding to anonymous functions defined
+	// inside inlined functions. These have names like:
+	// runtime.gcMarkDone.forEachP.func5
+	// "forEachP.func5" is an anonymous function defined inside forEachP, and
+	// "runtime.gcMarkDone" is the functions inside which "forEachP" is inlined.
+	parseAnonFuncInsideInlinedFunc = regexp.MustCompile(`^((?P<pkg>(.*/)?\w*?)\.)\w+\.\w+\.func(\d)+$`)
+)
+
+type FuncName struct {
+	Package string
+	// Type is the type of the receiver, if any. Empty if this function is not a
+	// method. The type is not a pointer type even if the method has a
+	// pointer-receiver; the base type is returned, without the '*'.
+	Type string
+	Name string
+	// QualifiedName looks like
+	// github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
+	QualifiedName string
+
+	// GenericFunction is set if the function takes type arguments. We don't
+	// support parsing these functions at the moment, so no other fields except
+	// QualifiedName are set.
+	GenericFunction bool
+}
+
+// ParseFuncName parses a Go qualified function name. For a qualifiedName name
+// like:
+// github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
+// the package is: github.com/cockroachdb/cockroach/pkg/kv/kvserver
+// the type is: raftSchedulerShard (note that it doesn't include the '*' signifying a pointer receiver).
+// the name is: worker
+//
+// This function should be kept in sync with parse_func_name in Rust.
+//
+// Cases we need to support:
+// github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker
+// github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed.rangefeedFactory.Run
+// github.com/klauspost/compress/zstd.sequenceDecs_decode_amd64
+// indexbytebody
+// github.com/.../pkg.myFunc.func0
+// internal/bytealg.init.0
```

Comment:
what about the deeper nested functions? their deferwraps?

`github.com/cockroachdb/cockroach/pkg/server.(*apiV2Server).execSQL.func8.1.3.2`
`github.com/.../logical.(*logicalReplicationWriterProcessor).flushBuffer.Group.GoCtx.func7.1.deferwrap1`

Look at this one: 
`github.com/cockroachdb/cockroach/pkg/server.(*topLevelServer).startPersistingHLCUpperBound.func1.(*Node).SetHLCUpperBound.1` ([code](https://github.com/cockroachdb/cockroach/blob/6bb7514c9db7662f51d37fb12935bd2d8735f3a3/pkg/server/clock_monotonicity.go#L243-L245))

Even worse, sometimes there's nested `func` labels:
`github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamproducer.(*eventStream).startStreamProcessor.(*eventStream).startStreamProcessor.func1.func3.1`